### PR TITLE
FIX  #172

### LINKED ISSUE #172: Invalid cos-agent relation data raises instead of blocking
### Bug Description

The new version of cos-agent lib switched to a new schema + pydantic validation.
As a result, charm code raises and the unit is in error state.
Instead, it should probably block.

### To Reproduce

Relate gagent to zookeeper (revision 96).

### Environment

```
Model                    Controller  Cloud/Region         Version  SLA          Timestamp
test-machine-agent-2za1  lxd         localhost/localhost  2.9.42   unsupported  10:29:04-04:00

App                  Version  Status  Scale  Charm          Channel        Rev  Exposed  Message
agent                         error       4  grafana-agent  edge             7  no       hook failed: "cos-agent-relation-joined"
principal-cos-agent           active      2  zookeeper      edge            96  no       
principal-juju-info  22.04    active      2  ubuntu         latest/stable   22  no       
```

### Relevant log output

```shell
unit-agent-2: 10:23:10.930 DEBUG unit.agent/2.juju-log cos-agent:4: Emitting Juju event cos_agent_relation_joined.
unit-agent-2: 10:23:11.028 ERROR unit.agent/2.juju-log cos-agent:4: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-agent-2/charm/./src/charm.py", line 473, in <module>
    main(GrafanaAgentMachineCharm)
  File "/var/lib/juju/agents/unit-agent-2/charm/venv/ops/main.py", line 441, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-agent-2/charm/venv/ops/main.py", line 149, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-agent-2/charm/venv/ops/framework.py", line 354, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-agent-2/charm/venv/ops/framework.py", line 830, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-agent-2/charm/venv/ops/framework.py", line 919, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-agent-2/charm/lib/charms/grafana_agent/v0/cos_agent.py", line 477, in _on_relation_data_changed
    provider_data = CosAgentProviderUnitData(**json.loads(raw))
  File "/var/lib/juju/agents/unit-agent-2/charm/venv/pydantic/main.py", line 341, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 5 validation errors for CosAgentProviderUnitData
metrics_alert_rules
  field required (type=value_error.missing)
log_alert_rules
  field required (type=value_error.missing)
dashboards
  field required (type=value_error.missing)
metrics_scrape_jobs
  field required (type=value_error.missing)
log_slots
  field required (type=value_error.missing)
unit-agent-2: 10:23:11.251 ERROR juju.worker.uniter.operation hook "cos-agent-relation-joined" (via hook dispatching script: dispatch) failed: exit status 1
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Abuelodelanada**
## Issue

This PR fixes #172

### CODE REVIEW

**Review by sed-i — APPROVED** [review]

File: tests/unit/machine/test_relation_status.py [hunk 1]
```diff
@@ -63,6 +64,12 @@ def test_cos_agent_with_relations(self):
         # THEN the charm goes into active status
         self.assertIsInstance(self.harness.charm.unit.status, ActiveStatus)
 
+    def test_cos_agent_wrong_provider_data(self):
+        rel_id = self.harness.add_relation("cos-agent", "principal")
+        self.harness.add_relation_unit(rel_id, "principal/0")
+        self.harness.update_relation_data(rel_id, "principal/0", {"config": "wrong_value"})
+        self.assertIsInstance(self.harness.charm.unit.status, BlockedStatus)
```
[the remark this answers is not in the document]
Comment (Abuelodelanada) [reply tests/unit/machine/test_relation_status.py]: Replaced by an scenario test

File: tests/scenario/test_machine_charm/test_cos_agent_e2e.py [hunk 2]
```diff
@@ -77,7 +77,16 @@ def vroot(placeholder_cfg_path):
...
+@pytest.fixture
+def snap_is_installed(autouse=True):
+    with patch(
+        "machine_charm.GrafanaAgentMachineCharm._is_installed", new_callable=PropertyMock
+    ) as mock_foo:
+        mock_foo.return_value = True
+        yield
+
+
+def test_cos_agent_e2e(vroot, snap_is_installed):
```
Comment (PietroPasotti) [inline tests/scenario/test_machine_charm/test_cos_agent_e2e.py:89]: **change requested** this test is huge and tests too many things. 
We should split every step out to a separate test.

File: tests/scenario/test_machine_charm/test_cos_agent_e2e.py [hunk 3]
```diff
@@ -94,50 +103,119 @@ def __init__(self, framework: Framework):
...
+
+    # Step 2: gagent is related to principal charm and ends in "blocked" status
+    # since there are missing relations:
+    #  - send-remote-write
+    #  - logging-consumer
+    #  - grafana-dashboards-provider
     peer = PeerRelation("peers")
     cos_agent1 = SubordinateRelation(
         "cos-agent",
         primary_app_name="mock-principal",
```
Comment (PietroPasotti) [inline tests/scenario/test_machine_charm/test_cos_agent_e2e.py:127]: **change** if you upgrade to scenario 3.0.2 you'll see that `primary-app-name` becomes `remote-app-name`
also if it's not pinned already, please pin the scenario version in tox/pyproject.toml

File: tests/scenario/test_machine_charm/test_cos_agent_e2e.py [hunk 4]
```diff
@@ -94,50 +103,119 @@ def __init__(self, framework: Framework):
...
         remote_unit_data=state_out.relations[0].local_unit_data,
     )
-    state_out1 = State(relations=[cos_agent1, peer]).trigger(
-        cos_agent1.changed_event(remote_unit_id=0),
+
+    ctx1 = Context(
```
Comment (PietroPasotti) [inline tests/scenario/test_machine_charm/test_cos_agent_e2e.py:131]: you shouldn't need to reinstantiate the context every time. In fact I typically add a single `ctx` fixture at the top of the test file (or even in a `conftest.py`) and reuse it throughout.

File: tests/scenario/test_machine_charm/test_cos_agent_e2e.py [hunk 5]
```diff
@@ -94,50 +103,119 @@ def __init__(self, framework: Framework):
...
+    state3 = State(leader=True, relations=[cos_agent2, peer_out, prometheus, loki])
+    state_out3 = ctx2.run(peer_out.changed_event(remote_unit_id=0), state=state3)
+    assert state_out3.status.unit.name == "blocked"
+
+    # Step 5: gagent is related to Grafana through "grafana-dashboards-provider" relation and ends
+    # in "active" status
+    grafana = Relation("grafana-dashboards-provider", remote_app_name="grafana-k8s")
+    state4 = State(leader=True, relations=[cos_agent2, peer_out, prometheus, loki, grafana])
+    state_out4 = ctx2.run(peer_out.changed_event(remote_unit_id=0), state=state4)
+    assert state_out4.status.unit.name == "active"
```
Comment (PietroPasotti) [inline tests/scenario/test_machine_charm/test_cos_agent_e2e.py:176]: **fyi** you can natively compare statuses: `assert state.status == ActiveStatus('foo')`

File: tests/scenario/test_machine_charm/test_cos_agent_e2e.py [hunk 6]
```diff
@@ -94,50 +103,119 @@ def __init__(self, framework: Framework):
...
+
+
+def test_cos_agent_wrong_rel_data(vroot, snap_is_installed):
+    class MyPrincipal(CharmBase):
```
Comment (PietroPasotti) [inline tests/scenario/test_machine_charm/test_cos_agent_e2e.py:180]: **change requested** I'd pull this out in a fixture and reuse it after you break this test into smaller ones

File: tests/scenario/test_machine_charm/test_peer_relation.py [hunk 7]
```diff
@@ -77,12 +77,12 @@ def test_no_dashboards():
     def post_event(charm: MyRequirerCharm):
         assert not charm.cosagent.dashboards
 
-    state.trigger(
+    ctx = Context(
```
Comment (PietroPasotti) [inline tests/scenario/test_machine_charm/test_peer_relation.py:80]: **repeated** pull ctx out in a fixture

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
main issue I see with the tests is that we should attempt centralizing our Contexts. 
The rest seems fine.
Good job! :)

File: tests/scenario/test_machine_charm/test_cos_agent_e2e.py (hunk 3 above)
Comment (Abuelodelanada) [reply tests/scenario/test_machine_charm/test_cos_agent_e2e.py:127]: I'm migrating to 3.x.. and if it's in 3.0.2 I'm ok with using `remote-app-name`... WDYT?

**Comment by Abuelodelanada** [comment]
Now that @sed-i fixed a lot of static issues in #196, we can close this PR in favour of  #198

**Pull request closed without merging**